File: frontend/tests/fixtures/analytics.ts
/**
 * Payloads shaped like the server's, shared by the five page tests.
 *
 * One module rather than five copies: the pages read different slices of the same
 * response, and five drifting fixtures would let a page pass against a shape the
 * server never sends.
 *
 * **Outside `src/` on purpose.** These are the only invented figures in the
 * project, and an audit asking "does anything ship fake data" should be able to
 * answer it from the directory layout rather than by tracing imports. The eslint
 * config forbids importing this directory from anything that is not a test.
 */

import { vi } from 'vitest';

import type {
  AnalyticsInsights,
  NamedCount,
  PerformancePage,
  PerformanceRow,
  RankedSku,
  SkuStatus,
} from '../../src/types/api';

export type Route = { ok: boolean; status: number; body: unknown };

export function ranked(sku: string, over: Partial<RankedSku> = {}): RankedSku {
  return {
    rank: 1,
    sku,
    sku_normalized: sku.replace(/-/g, '').toLowerCase(),
    shopify_sales: 0,
    shopify_sales_pct: 0,
    total_complaints: 0,
    total_qty: 0,
    total_orders: 0,
    ...over,
  };
}

function count(label: string, value: number, share: number): NamedCount {
  return {
    field_name: label.toLowerCase().replace(/ /g, '_'),
    label,
    count: value,
    share_pct: share,
  };
}

export const COMPLAINT_COLUMNS = [
  { field: 'item_defect_partial', header: 'Item Defect Partial' },
  { field: 'item_damage_partial', header: 'Item Damage Partial' },
];

export const TREND = {
  points: [10, 20, 30].map((units, i) => ({
    day: `2026-07-2${String(i + 6)}`,
    units,
    revenue_paise: 0,
  })),
  previous: [5, 5, 5].map((units, i) => ({
    day: `2026-07-2${String(i + 3)}`,
    units,
    revenue_paise: 0,
  })),
  days: 30,
};

/**
 * Every complaint in these fixtures is dated, so nothing is unfilterable and no
 * note is shown. The undated case has its own fixture below, because that is a
 * different screen rather than a different number.
 */
export const DATED_SCOPE = {
  filtered_by_date: true,
  dated_skus: 4,
  undated_skus: 0,
  undated_complaints: 0,
};

/** An aggregated sheet: no Complaint Date column anywhere in the workspace. */
export const UNDATED_SCOPE = {
  filtered_by_date: false,
  dated_skus: 0,
  undated_skus: 4,
  undated_complaints: 33,
};

/**
 * The common case once a store has imported twice: an aggregated sheet, then a
 * dated export covering part of it. Modelled on the live workspace, where 308
 * SKUs carried dates and 883 did not.
 */
export const MIXED_SCOPE = {
  filtered_by_date: true,
  dated_skus: 308,
  undated_skus: 883,
  undated_complaints: 5456,
};

export const INSIGHTS: AnalyticsInsights = {
  complaint_scope: DATED_SCOPE,
  kpis: {
    total_skus: 4,
    total_qty: 1050,
    shopify_sales: 325,
    shopify_sales_pct: 76.5,
    total_orders: 340,
    total_complaints: 33,
    avg_sales_per_sku: 81.2,
    shopify_sales_all: 425,
  },
  sales: {
    shopify_sales: 325,
    shopify_sales_pct: 76.5,
    highest: ranked('DD-1001', { shopify_sales: 300, shopify_sales_pct: 70.59 }),
    lowest: ranked('DD-1003', { rank: 4 }),
    top: [
      ranked('DD-1001', { rank: 1, shopify_sales: 300, shopify_sales_pct: 70.59 }),
      ranked('DD-1002', { rank: 2, shopify_sales: 20, shopify_sales_pct: 4.71 }),
    ],
    distribution: [count('DD-1001', 300, 92.3), count('DD-1002', 20, 6.2)],
  },
  complaints: {
    total_complaints: 33,
    most_complained: ranked('DD-1002', { total_complaints: 20 }),
    categories: [
      count('Item Defect Partial', 21, 63.6),
      count('Item Damage Partial', 12, 36.4),
    ],
    top_skus: [
      ranked('DD-1002', {
        rank: 1,
        total_complaints: 20,
        shopify_sales: 40,
        total_orders: 40,
      }),
    ],
    skus_with_complaints: 3,
  },
  rankings: {
    top_selling: [ranked('DD-1001', { rank: 1, shopify_sales: 300, shopify_sales_pct: 70.6 })],
    lowest_selling: [ranked('DD-1003', { rank: 1, total_qty: 900 })],
    highest_complaint: [
      ranked('DD-1002', {
        rank: 1,
        total_complaints: 20,
        total_orders: 40,
      }),
    ],
  },
  inventory: {
    high_stock_low_sales: [ranked('DD-1003', { rank: 1, total_qty: 900 })],
    low_stock_high_sales: [ranked('DD-1002', { rank: 1, total_qty: 40, shopify_sales: 20 })],
    zero_sales: [ranked('DD-1003', { rank: 1, total_qty: 900 })],
    most_complaints: [
      ranked('DD-1002', {
        rank: 1,
        total_complaints: 20,
        total_orders: 40,
      }),
    ],
    median_qty: 70,
    median_sales: 12.5,
    zero_sales_total: 1,
  },
  quick: [
    {
      key: 'best',
      icon: 'check',
      title: 'Best performing SKU',
      sku: 'DD-1001',
      value: '300 units',
      note: 'highest sales with no complaints logged',
    },
    {
      key: 'restock',
      icon: 'bell',
      title: 'Restock needed',
      sku: 'DD-1002',
      value: '40 left',
      note: '300 sold in the window',
    },
    {
      key: 'overstocked',
      icon: 'box',
      title: 'Overstocked SKU',
      sku: 'DD-1003',
      value: '900 units held',
      note: 'only 0 sold',
    },
    {
      key: 'nosales',
      icon: 'x',
      title: 'No sales',
      sku: 'DD-1003',
      value: '1 SKUs',
      note: 'holding stock but unsold — DD-1003 has the most at 900',
    },
  ],
  trend: TREND,
  complaint_columns: COMPLAINT_COLUMNS,
  days: 30,
  has_data: true,
  stale: false,
  syncing: false,
  last_computed_at: '2026-07-30T09:00:00Z',
};

function row(
  sku: string,
  status: SkuStatus,
  over: Partial<PerformanceRow> = {},
): PerformanceRow {
  return {
    sku,
    sku_normalized: sku.replace(/-/g, '').toLowerCase(),
    total_count: 120,
    total_qty: 10,
    total_orders: 100,
    shopify_sales: 50,
    shopify_sales_pct: 11.76,
    total_complaints: 0,
    status,
    complaints: {},
    ...over,
  };
}

export const PERFORMANCE: PerformancePage = {
  complaint_scope: DATED_SCOPE,
  rows: [
    row('DD-1001', 'excellent', {
      total_qty: 100,
      shopify_sales: 300,
      shopify_sales_pct: 70.59,
    }),
    row('DD-1002', 'critical', { total_qty: 40, total_complaints: 20 }),
    row('DD-1003', 'attention', { total_qty: 900, shopify_sales: 0, total_orders: 0 }),
  ],
  complaint_columns: COMPLAINT_COLUMNS,
  total: 3,
  limit: 50,
  offset: 0,
  days: 30,
  sort: 'shopify_sales',
  descending: true,
};

export function routes(overrides: Record<string, Route> = {}) {
  const table: Record<string, Route> = {
    'GET /analytics/insights': { ok: true, status: 200, body: INSIGHTS },
    'GET /analytics/performance': { ok: true, status: 200, body: PERFORMANCE },
    'POST /analytics/rebuild': {
      ok: true,
      status: 200,
      body: { rows_written: 12, days_covered: 30, duration_ms: 40 },
    },
    ...overrides,
  };
  return vi.fn((input: RequestInfo | URL, init?: RequestInit) => {
    const url = String(input);
    const method = (init?.method ?? 'GET').toUpperCase();
    const key = Object.keys(table)
      .sort((a, b) => b.length - a.length)
      .find((candidate) => {
        const [routeMethod = '', routePath = ''] = candidate.split(' ');
        return routeMethod === method && url.includes(routePath);
      });
    const route = (key ? table[key] : undefined) ?? { ok: true, status: 200, body: {} };
    return Promise.resolve({
      ok: route.ok,
      status: route.status,
      json: () => Promise.resolve(route.body),
    } as Response);
  });
}

/** An insights response with one branch replaced. */
export function insightsWith(patch: Partial<AnalyticsInsights>) {
  return routes({
    'GET /analytics/insights': { ok: true, status: 200, body: { ...INSIGHTS, ...patch } },
  });
}

/** The URLs a page requested, in order. */
export function requested(fetch: ReturnType<typeof routes>, path: string): string[] {
  return fetch.mock.calls.map((call) => String(call[0])).filter((url) => url.includes(path));
}

File: backend/app/schemas/analytics.py
"""Request and response bodies for analytics.

Six cards and one table. The uploaded sheet is the source of truth for
everything except ``shopify_sales`` and ``shopify_sales_pct``, which Shopify
supplies and which are joined to a sheet row on the normalised SKU alone.
"""

from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel

#: The compact per-chart range control (design doc §7.3).
RANGE_PATTERN = "^(7|30|90|180|365)$"
#: The stock badge on a table row, derived from Quantity against the workspace
#: threshold. There is no Shopify inventory feed left to disagree with it.


class ComplaintScopePayload(BaseModel):
    """Whether the complaint figures on this page follow the selected range.

    Sent with every payload that carries a complaint number, because the number
    alone cannot say: "no complaints this month" and "complaints we cannot place
    in a month" are the same figure on screen.

    Counts, not copy. The sentence is the client's, built from these three in
    `ComplaintScopeNote`, where the numbers can be grouped for the locale.
    """

    #: True when at least some SKUs were imported with a Complaint Date column.
    filtered_by_date: bool = False
    dated_skus: int = 0
    undated_skus: int = 0
    #: Complaints no range can hold. The count is data, so a page can say how
    #: much of the tally it is not filtering.
    undated_complaints: int = 0


class KpiPayload(BaseModel):
    """The six dashboard cards, plus the context the percentage needs."""

    # From the uploaded sheet.
    total_skus: int
    total_quantity: int
    total_orders: int
    total_complaints: int
    #: Whether the figure above followed the range, and what to say if not.
    complaint_scope: ComplaintScopePayload = ComplaintScopePayload()

    # From Shopify, matched by SKU.
    shopify_sales: int
    #: Share of all Shopify units belonging to a SKU in the sheet. Bounded 0–100.
    shopify_sales_pct: float

    #: The denominator behind ``shopify_sales_pct`` — every unit the store sold in
    #: the window, matched or not. Returned so the card can be explained rather
    #: than taken on trust.
    shopify_sales_all: int
    revenue_paise: int
    low_stock: int
    #: Returned so the UI can say *why* a SKU counts as low without hard-coding a
    #: number that lives on the workspace.
    low_stock_threshold: int

    days: int
    #: True when orders exist that the rollup has not seen and nothing is
    #: working on it — which now means the automatic recompute failed.
    stale: bool
    #: True while a sync is queued or running. Mutually exclusive with `stale`
    #: by construction, so a page shows one message or the other and never both:
    #: figures being rebuilt are not figures that are behind.
    syncing: bool = False
    last_computed_at: datetime | None


class TrendPointPayload(BaseModel):
    day: date
    units: int
    revenue_paise: int


class TrendPayload(BaseModel):
    points: list[TrendPointPayload]
    #: The preceding window of equal length, for the comparison line.
    previous: list[TrendPointPayload]
    days: int


class ComplaintColumn(BaseModel):
    """One complaint category, so the client renders columns from the server."""

    field: str
    header: str


class SkuRowPayload(BaseModel):
    sku: str
    sku_normalized: str
    quantity: int
    shopify_sales: int
    shopify_sales_pct: float
    total_orders: int
    total_qty: int
    total_count: int
    #: Keyed by the same ``field`` values ``complaint_columns`` carries.
    complaints: dict[str, int]
    total_complaints: int
    stock_status: str


class SkuTablePage(BaseModel):
    rows: list[SkuRowPayload]
    #: The complaint columns, in sheet order. Sent with the page so the table's
    #: headers and its cells cannot disagree about the set.
    complaint_columns: list[ComplaintColumn]
    complaint_scope: ComplaintScopePayload = ComplaintScopePayload()
    total: int
    limit: int
    offset: int
    days: int


# ---------------------------------------------------------------------------
# Analytics page (business intelligence). The dashboard's payloads above stay
# as they are: the two pages answer different questions and must be free to
# change independently.
# ---------------------------------------------------------------------------

#: A SKU's automatic verdict, worst to best.
STATUS_PATTERN = "^(critical|attention|good|excellent)$"
#: Columns the performance table may be sorted by.
SORT_PATTERN = (
    "^(sku|total_count|total_qty|total_orders|shopify_sales|shopify_sales_pct"
    "|total_complaints|status)$"
)


class InsightKpiPayload(BaseModel):
    """The eight cards at the top of Analytics."""

    total_skus: int
    total_qty: int
    shopify_sales: int
    shopify_sales_pct: float
    total_orders: int
    total_complaints: int
    #: Units sold per SKU carried.
    avg_sales_per_sku: float
    shopify_sales_all: int


class RankedSkuPayload(BaseModel):
    rank: int
    sku: str
    sku_normalized: str
    shopify_sales: int
    shopify_sales_pct: float
    total_complaints: int
    total_qty: int
    total_orders: int


class NamedCountPayload(BaseModel):
    """One slice or bar. ``field_name`` is the complaint attribute or a SKU."""

    field_name: str
    label: str
    count: int
    share_pct: float


class SalesAnalyticsPayload(BaseModel):
    shopify_sales: int
    shopify_sales_pct: float
    #: Absent when nothing sold, rather than a zero row that reads as a finding.
    highest: RankedSkuPayload | None
    lowest: RankedSkuPayload | None
    top: list[RankedSkuPayload]
    distribution: list[NamedCountPayload]


class ComplaintAnalyticsPayload(BaseModel):
    total_complaints: int
    most_complained: RankedSkuPayload | None
    categories: list[NamedCountPayload]
    top_skus: list[RankedSkuPayload]
    skus_with_complaints: int


class RankingsPayload(BaseModel):
    top_selling: list[RankedSkuPayload]
    lowest_selling: list[RankedSkuPayload]
    highest_complaint: list[RankedSkuPayload]


class InventoryInsightsPayload(BaseModel):
    high_stock_low_sales: list[RankedSkuPayload]
    low_stock_high_sales: list[RankedSkuPayload]
    zero_sales: list[RankedSkuPayload]
    most_complaints: list[RankedSkuPayload]
    #: The cuts the lists were made at, so the UI states them rather than
    #: leaving "high" and "low" to the reader's imagination.
    median_qty: float
    median_sales: float
    zero_sales_total: int


class QuickInsightPayload(BaseModel):
    key: str
    #: Names an icon the client already has.
    icon: str
    title: str
    sku: str | None
    value: str
    note: str


class AnalyticsInsights(BaseModel):
    """Everything on Analytics except the performance table, in one request.

    One call rather than eight: the figures come from a single read of the SKU
    facts, and splitting them across requests would let two panels on the same
    screen describe two different moments.
    """

    kpis: InsightKpiPayload
    sales: SalesAnalyticsPayload
    complaints: ComplaintAnalyticsPayload
    rankings: RankingsPayload
    inventory: InventoryInsightsPayload
    quick: list[QuickInsightPayload]
    trend: TrendPayload
    complaint_columns: list[ComplaintColumn]
    complaint_scope: ComplaintScopePayload = ComplaintScopePayload()
    days: int
    has_data: bool
    #: True when orders exist that the rollup has not seen and nothing is
    #: working on it — which now means the automatic recompute failed.
    stale: bool
    #: True while a sync is queued or running. Mutually exclusive with `stale`
    #: by construction, so a page shows one message or the other and never both:
    #: figures being rebuilt are not figures that are behind.
    syncing: bool = False
    last_computed_at: datetime | None


class PerformanceRowPayload(BaseModel):
    sku: str
    sku_normalized: str
    total_count: int
    total_qty: int
    total_orders: int
    shopify_sales: int
    shopify_sales_pct: float
    total_complaints: int
    status: str
    complaints: dict[str, int]


class PerformancePage(BaseModel):
    rows: list[PerformanceRowPayload]
    complaint_columns: list[ComplaintColumn]
    complaint_scope: ComplaintScopePayload = ComplaintScopePayload()
    total: int
    limit: int
    offset: int
    days: int
    sort: str
    descending: bool


class RebuildResultPayload(BaseModel):
    rows_written: int
    days_covered: int
    duration_ms: int


class AnalyticsOverview(BaseModel):
    """Everything the dashboard needs above the table, in one request."""

    kpis: KpiPayload
    trend: TrendPayload
    has_data: bool
